### DISCUSSION TITLE
blog/component-composition-is-great-btw

### DISCUSSION THREAD

**Discussion by giscus** [TkDodo/blog-comments#110] [Announcements]
# blog/component-composition-is-great-btw

Component composition is one of the best parts of React, and I think we should take more time to break our components into manageable parts before littering one component with conditional renderings.

https://tkdodo.eu/blog/component-composition-is-great-btw

**Comment by palaniichukdmytro** [comment]
Great one , I think there is more topics were described , in scope of RSC it means even more sense to keep data to ui , early return , self contained . I think minor update might be in a code

```
   {title &&  <CardHeading>Welcome 👋 {title}</CardHeading>}
```

**Reply by TkDodo** [reply]
This would not render the heading at all in the branches where we don't pass a title. It's fine in react to just render `undefined`

**Reply by ryan-zayne** [reply]
But then wouldn't you just have a useless Dom node sitting in the tree?

**Reply by ryan-zayne** [reply]
Due to the CardHearding wrapper?

**Reply by TkDodo** [reply]
My example always renders the `Welcome 👋` message so it isn't unnecessary

**Reply by ryan-zayne** [reply]
Makes sense👀

**Comment by CodeWithRomaen** [comment]
Always love reading your posts! Fantastic examples and really easy to follow. Excited to try this pattern

**Reply by thepradipvc** [reply]
💯 True
Directly jumped to it as soon as I saw the twitter Post.

**Comment by Deer404** [comment]
Very good component packaging techniques, I always think that don't rush to write code, first think!

**Comment by iam-hbk** [comment]
Was waiting for it after that X thread 👌🏽

**Comment by jbee37142** [comment]
Great post, I really enjoyed reading it. As someone who often thinks about good component design, I found many parts relatable. I'd like to share a couple of thoughts that came to mind.

**1. Error Handling**
I see that loading is covered, but when handling asynchronous tasks, it's also important to account for possible failures (errors), much like how `Promises` work. When designing 'self-contained' components that handle their own data fetching, I believe we should consider three states:

- pending
- resolved
- rejected

It would be great to include a branch for error handling as well. I’ve written a post on handling client-side errors that might be relevant, so I’ll leave it here for reference: [Client-Centered User Exception Handling](https://dev.to/jaeyeophan/client-centered-user-exception-handling-23a1
).

**2. Separation of Concerns**
Next, I think it could be useful to separate what's essential from what’s less critical. For example, using Suspense and ErrorBoundary as shown in your example might help reduce the complexity of conditional logic inside the component.
Considering the Promise states mentioned above, we could handle them like this:

- pending -> Suspense fallback component
- resolved -> Main component (empty view, list view, etc.)
- rejected -> ErrorBoundary fallback component

Thanks again for the great post! It gave me a lot to think about, and I’m looking forward to applying some of these ideas in my own work.

**Comment by andrei9669** [comment]
for pure components, this would work really good but for stateful components this feels like it falls apart as soon as you need anything more complex. fair enough, you could have a separate layout for each state but that now sounds like lots of code duplication.

**Reply by igortas** [reply]
> for pure components, this would work really good but for stateful components this feels like it falls apart as soon as you need anything more complex. fair enough, you could have a separate layout for each state but that now sounds like lots of code duplication.

love how all other replies gets likes, but real honest comment, is bypassed. I'm on your track, React is just not scalable option for big projects. Composition or not, Layout wrapper and pass components through props of the layout, everything fall apart.

**Reply by TkDodo** [reply]
can you show an example where things fall apart? Really hard to discuss in the abstract :)

**Reply by igortas** [reply]
> can you show an example where things fall apart? Really hard to discuss in the abstract :)

@TkDodo
React fall short, when u have so much logic, that needs to be orchestrated, every click change something near or far, and few more optimizations tricks for rerendering etc, won't save u on long run, because u will get shoot in the foot, and at the end u just let everything rerender like wild. At least this is my experience on my current struggles with project with many and tight connected logic.
How much more I'm thinking about FE, the more I see CQRS the only way, just dispatch everything on the BE, and don't worry about anything, or pub/sub pattern (where u really don't need any framework)...

**Reply by PieterDePauw** [reply]
You are wrong. If every click in your front-end truly leads to side effects, the declarative approach of React is just not the right tool to achieve those requirements. In most cases, you shouldn't have to worry too much about multiple re-renderings of your components; that shouldn't be the problem. However, it might also just mean that you are either (1) not really closely following, or at worst, even working against the mental model of React, being "UI = f(state)" or (2) offloading too much slow and expensive operations to the client that better would be solved on the server / on the back-end.

React is very scalable, however most developers are not and try to just hack some things together.

**Comment by gpichot** [comment]
Interesting, I went all the way and created a pattern matching function for this:

It works like this:
```typescript
export default function SessionsList() {
  const sessionsQuery = useSessionsListQuery();

  return (
    <PageLayout title="My sessions">
      {matchQueryStatus(sessionsQuery, {
        Loading: <Skeleton />
        Errored: <Alert kind="error" icon={<IconAlertCircle size="1rem" />} />,
        Empty: <Alert kind="default" description="No session scheduled" />,
        Success: ({ data }) => <SessionList items={data} />
      })}
    </PageLayout>
  );
}
```

I spent a bit of time so that the function is a bit "smart":
- if there is no Empty tag, data in the Success callback will be null or undefined 
- if there is an Empty, data in the Success callback with be non nullish so in a way it forces me to add the Empty tag.

It saves me a lot of time and is clearer than conditionals.

I created a gist here https://gist.github.com/gpichot/df6b586c0ac09ebde794c7f5bef5800f with the corresponding function.

**Reply by ikovic** [reply]
This is a really nice pattern, thanks for sharing!
We do something quite similar whenever the states the component can be in are mutually exclusive, this does wonders for readability & essentially writes the unit test scenarios by itself.

**Reply by gpichot** [reply]
Yes I use this a lot, most of the time it removes half of the code :)
But as everything it's not a silver bullet. It matches 90% of my use cases though.

**Reply by igortas** [reply]
@gpichot
This is nothing different, than attaching thousand of props on the Layout component as child nodes, which still u have bunch of noise below, thousand of useQuery calls...

**Reply by Tungetyt** [reply]
You are going in right direction, but there is a lot to be improved in order for this function to be actually useful. You need to cover cases like:
1. cache data exists, but refetch errored out. You should probably display cached data with error alert.
2. you should handle isFetching too, not only isLoading.

**Reply by gpichot** [reply]
@igortas This is completely different.

@Tungetyt It covers all the cases that I have and it's actually pretty useful. You may have missed it, but you can extract `isFetching` and `isRefetchError` from the Success parameter. :-)

**Reply by igortas** [reply]
@gpichot can u tell me what is the different, because Success gets at the end the list to present, and if u have complex tree components structure if that is how u structure the code, u get far worse rerendering than puting props on Layout component.

**Reply by k-sav** [reply]
Nice. Kind of reminds me of the [Redwoodjs concept of a cell](https://docs.redwoodjs.com/docs/tutorial/chapter2/cells)

**Comment by harshtalks** [comment]
One thing I always prefer in such scenario is a pattern matching. For a strongly typed libraries such as react-query, pattern-matching brings more simplicity to the codebase.

**Reply by seanblonien** [reply]
Can you give an example of pattern patching in JSX? I am unfamiliar with the concept and what to understand what you mean

**Reply by ryan-zayne** [reply]
> Can you give an example of pattern patching in JSX? I am unfamiliar with the concept and what to understand what you mean

Check out solidjs' Switch component for starters

**Comment by burdiuz** [comment]
I would probably make an enum of states , used switch and extracted it from the component logic into separate function. There would probably be another function to identify current state, extenal to the component.

**Comment by wearypossum4770** [comment]
One thing I like to do is use a swith statement to render components.

**Comment by Melh1or** [comment]
Great article, I really like what you are doing.
What other drawbacks can you mention in cases when we want to eliminate layout duplication?
For example, if I write something like this:
```tsx
function Layout(props: { children: ReactNode; title?: string }) {
  return (
    <Card>
      <CardHeading>Welcome 👋 {title}</CardHeading>
      <CardContent>{props.children}</CardContent>
    </Card>
  )
}

export function ShoppingList() {
  const { data, isPending } = useQuery(/* ... */)

  const content = useMemo(() => {
    if (isPending) {
      return <Skeleton />
    }

    if (!data) {
      return <EmptyScreen />
    }

    return (
      <>
        {data.assignee ? <UserInfo {...data.assignee} /> : null}
        {data.content.map((item) => (
          <ShoppingItem key={item.id} {...item} />
        ))}
      </>
    )
  }, [data, isPending])

  return <Layout title={data.title}>{content}</Layout>
}
```
we can also return not only the `content`, but also an object with a `title` and `content`

**Reply by TkDodo** [reply]
This doesn’t evolve that well, because we re-add complexity. For example, `title={data.title}` won’t work because `data` can be `undefined` now, so you need at least `data?.title`. But data can, again, be `undefined` for various reasons. So we are again adding conditions and branches to the “generic” branch that we would like to keep straight-forward and top-down. And in my experience, we will sooner or later wind up again with something like:

```
return <Layout title={data?.title} showMoreInfo={!isPending && data && data.content.length > 0 }>{content}</Layout>
```

or whatever crappy condition we can imagine 😂 

And yeah at that point, it’s just way easier to have the call to `Layout` duplicated among 3 conditions and just inline the condition in the one place where it’s relevant.

**Comment by SviatoslavChyzh** [comment]
in line 4 of last example it should be {props.title} instead of {title} only

**Reply by TkDodo** [reply]
fixed 🎉

**Comment by LovesWorking** [comment]
I really like this! I'll make sure to use this pattern going forward.

I'm curious though - if we're implementing this across all our components, would you recommend using a reusable wrapper like this:

```jsx
function QueryWrapper({ isLoading, error, data, LoadingView, ErrorView, children }) {
  if (isLoading) return LoadingView || <DefaultSpinner />
  if (error) return ErrorView || <DefaultError error={error} />
  return children(data)
}
```
Or stick with the pattern you described with direct conditional returns? I would love to know which approach you prefer.

**Reply by wearypossum4770** [reply]
Looks good but that would be crazy to do in Typescript.

**Reply by LovesWorking** [reply]
> Looks good but that would be crazy to do in Typescript.

How so?  This is very simple?

**Reply by LovesWorking** [reply]
```jsx
import { UseQueryResult } from "@tanstack/react-query";

interface QueryStateHandlerProps {
  queryObj: UseQueryResult<any, Error>;
  children: React.ReactNode;
  errorMessage?: string;
  LoadingView?: React.ReactNode;
}

export default function QueryStateHandler({
  queryObj,
  children,
  errorMessage = "Failed to load data",
  LoadingView,
}: QueryStateHandlerProps) {
  if (queryObj.isLoading) {
    return LoadingView ?? <DefaultLoadingView />;
  }

  if (queryObj.isError) {
    return (
      <div>
        <p> error: {errorMessage}</p>
        <button onClick={() => queryObj.refetch()}>Retry</button>
      </div>
    );
  }

  return children;
}

```

use

```jsx
    <QueryStateHandler
        queryObj={analyticsQuery}
        LoadingView={<SplashScreen />}
        errorMessage="Failed to load data"
      >
        <Anything/>
      </QueryStateHandler>
```

**Reply by LovesWorking** [reply]
@wearypossum4770  @TkDodo see any issues with that? Would you recommend this? Thanks.

**Reply by TkDodo** [reply]
if a background query errors, you will see the error message instead of stale data. I don’t think this approach is worth it. I would rather just use react suspense, because it’s made for decoupling your component from loading and error states.

**Comment by ploca14** [comment]
What if the layout contains an input? Won't it loose focus when re-rendered in a different path?

**Reply by gpichot** [reply]
@ploca14 It will work, unless you have different elements "path".

This will work
```typescript
function MyComp() {
  if (isLoading) return <div><input {...} /></div>
  
  return <div><input {...} /></div>
}
```

This will not:
```typescript
function MyComp() {
  if (isLoading) return <div><input {...} /></div>
  
  return <input {...} />
}
```

**Reply by TkDodo** [reply]
re-rendering the same component at the same position (in the same component) with the same `key` will not result in state loss. So unless you pass a `key` to `Layout`, or you conditionally render the input at different positions within the `Layout`, state will be kept around.

**Comment by allpro** [comment]
Imo, the Layout should be a separate, parent component. It contains the logic to know which 'child component' to render. This is the most decoupled and modular pattern, with zero relation. 

The 3 child components should not know anything about each other, or the "Layout" they are inside. That is beyond the scope of their knowledge and responsibility. The role of the CardContent component is to render that content, and that's all!

If a different child of the Layout needs to render a "title", it should not be reliant on the 'sibling' CardContent component to pass it data. Again, neither of these sub-components should know anything about each other - they have separate roles.

In the example given, the "data" is a prop passed down to the CardContent. Therefore a parent Layout component would have the `data.title`, and all other information needed to CHOOSE which children to display, and the necessary props to pass. This is the Layout's role. Separation of concerns.

If the Layout JSX and/or data handling is quite complex, these things could be further decoupled by moving the data logic into a hook. Then the Layout component code can focus on the JSX, styles, etc. The hook handles the data logic and passes back what the Layout needs to know, like the 'state'.

The more decoupled components and data/logic are, the simpler, more versatile and more maintainable everything becomes. For example, the CardContent could be rendered in a 'container' other than the Card - like a modal or sidebar - because it is modular, ignorant of its parent, and purely prop-driven. This is a sign that a components is well decoupled. 

My 2-cents.

**Reply by TkDodo** [reply]
Seems like you prefer a smart/dumb component split, which is something I’m usually against. I’m touching on this topic here:

- https://tkdodo.eu/blog/react-query-as-a-state-manager#a-note-on-separation-of-concerns

and the linked talk by Mark Erikson is also a great watch.

**Comment by alexander-zuev** [comment]
Gems like these is what I am on Internet for
